### DISCUSSION TITLE
RFC: Drift postgres with postgres3 PgPool

### DISCUSSION THREAD

**Discussion by ignatz** [simolus3/drift#2569] [Q&A]
Hey Simon,

Firstly, :pray: for adding a PgPool implementation to postgres3. You seem to be everywhere.

Until know, I've been running a relatively old patched version of drift postgres to run with a postgres_pool. I was very excited to see your commit thinking that I could finally switch to a modern, upstream version.

Looking at it  a bit more closely the current design of drift_postgres' _PgDelegate being tight to singular "_openedSession" doesn't quite seem to fit the PgPool's on-demand withConnection model... or I am missing something :)

Is this something you're already working on or would you be interested in a PR?

**Comment by simolus3** [comment]
Thanks for the kind words! I think supporting pools would be great . `PgPool` implements `PgSession` already (by implicitly calling `withConnection` all the time), so doing this with `PgDatabase.opened` should already kind of work (let me know if it doesn't, I'll have to look into that then).

A PR to use pools more easily (perhaps a maximum connection count parameter on the default constructor) would be welcome of course.

**Reply by ignatz** [reply]
> Thanks for the kind words! I think supporting pools would be great . PgPool implements PgSession already (by implicitly calling withConnection all the time), so doing this with PgDatabase.opened should already kind of work (let me know if it doesn't, I'll have to look into that then).

Mea culpa, I should have mentioned. I tried and it didn't work for me. I assumed it was by design that it didn't implicitly call withConnection, but it seems like it should. Specifically, using

```dart
  final pgPool = PgPool(/**/);
  final db = MyDatabase(DatabaseConnection(PgDatabase.opened(pgPool)));
```

and then interacting with the `db` hangs, whereas hackery like:

```dart
  final pgPool = PgPool(/**/);
  final connection = await pgPool.run((session) async {
      return DatabaseConnection(PgDatabase.opened(session));
  });
  final db = MyDatabas(connection.executor);
```

works but clearly doesn't give me the pool properties and might even be racy.

I was looking into a change that would change _PgDelegate to explicitly call `withConnection` under the hood.

On a tangent and since I have your attention already, does the new PgPool implementation only manage static connections or does it also monitor the connections and re-establish when necessary?

**Reply by simolus3** [reply]
> I tried and it didn't work for me. I assumed it was by design that it didn't implicitly call withConnection, but it seems like it should

Hm yes, this sounds like a bug in the postgres pool implementation that I should look into.

> On a tangent and since I have your attention already, does the new PgPool implementation only manage static connections or does it also monitor the connections and re-establish when necessary?

At the moment it only manages them statically. Eventually we should look into monitoring connections for failures as well.

**Reply by ignatz** [reply]
let me know if there's anything I can do to help. Meanwhile, I'll halt any work on making the _PgDelegate call withConnection explicitly.

> At the moment it only manages them statically. Eventually we should look into monitoring connections for failures as well.

I see. I was using pools (https://pub.dev/packages/postgres_pool) mostly to automatically re-establish connections when the underlying postgres goes down. I don't expect you to be dart-postgres' coding monkey but maybe it would then make more sense for drift_postgres to support pre v3 dart-postgres and https://pub.dev/packages/postgres_pool? Is this something you'd be open to? I can take a look

**Reply by simolus3** [reply]
Trying to use the v2 postgres API for drift turned out to be a massive rabbit hole that lead to the v3 draft and my work on its implementation :)

The main problem is that we can't disable the built-in query transformation logic in the v2 API - for drift it makes sense to control variables explicitly though. I originally wanted to add an API for that to the v2 version, but that was rejected in favor of the v3 draft.
